server/src/main/java/org/tctalent/server/repository/db/CandidateRepository.java

```diff
@@ -688,4 +689,175 @@ List<Object[]> countByReferrerOrderByCount(@Param("gender") String gender,
+                JOIN JobChat jc ON c2.id = jc.candidate.id
+                WHERE jc.type = 'CandidateProspect'
+                AND jc.id IN (
+                    SELECT jc2.id
+                    FROM JobChat jc2
+                    JOIN ChatPost cp ON jc2.id = cp.jobChat.id
+                )
+            )
+            """
+    )
+    Page<Candidate> findCandidatesWithActiveChat(
+        @Param("partnerId") long partnerId,
+        @Param("keyword") String keyword,
+        Pageable pageable
+    );
+
+    /**
+     * Search finding IDs of candidates managed by the logged-in user's partner organisation, if they
+     * have a Job Chat of type 'CandidateProspect' containing at least one post that is unread by
+     * the logged-in user.
+     * Being a complex query, this was not suitable for JPQL so is instead written in regular SQL,
+     * requiring return of IDs rather than the whole candidate object.
+     * @param partnerId logged-in user's partner org's ID
+     * @param userId logged-in user's ID
+     * @param keyword pre-processed keyword entered in search filter, may represent candidate
+     *                number, first name or last name of candidate
+     * @return list of IDs of candidates who match the criteria
+     */
+    @Query(
+        value =
+            """
+            SELECT candidate.id
+            FROM candidate
+                     JOIN users ON candidate.user_id = users.id
+            WHERE candidate.id IN (
+                SELECT job_chat.candidate_id
+                FROM (
+                         SELECT job_chat.id
+                         FROM candidate
+                                  JOIN job_chat ON candidate.id = job_chat.candidate_id
+                         WHERE type = 'CandidateProspect'
+                     ) AS chats
+                         JOIN job_chat ON job_chat.id = chats.id
+                WHERE
+                    (SELECT last_read_post_id
+                     FROM job_chat_user
+                     WHERE job_chat_id = chats.id AND user_id = :userId) <
+                    (SELECT max(id)
+                     FROM chat_post
+                     WHERE job_chat_id = chats.id)
+                   OR (
+                    (SELECT last_read_post_id
+                     FROM job_chat_user
+                     WHERE job_chat_id = chats.id AND user_id = :userId) IS NULL
+                        AND
+                    (SELECT count(*)
+                     FROM chat_post
+                     WHERE job_chat_id = chats.id) > 0
+                    )
+            )
+              AND users.partner_id = :partnerId
+              AND (
+                LOWER(users.first_name) LIKE :keyword
+                OR LOWER(users.last_name) LIKE :keyword
+                OR candidate.candidate_number LIKE :keyword
+                )
+            """, nativeQuery = true
+    )
+    List<Long> findCandidatesWithActiveAndUnreadChat(
+        @Param("partnerId") long partnerId,
```

Comment:
Would be better to name this findCandidate_**Ids**_WithActive...

It is a bit confusing to have two very similarly named "findCandidates..." methods - one which returns candidates and the other returns candidate ids.

Reply:
Done 👍